testsuite/cluster-test/src/atomic_histogram.rs

```diff
@@ -0,0 +1,191 @@
+
+#![forbid(unsafe_code)]
+
+use std::{
+    fmt::Debug,
+    ops::Sub,
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+const DEFAULT_HISTOGRAM_CAPACITY: usize = 2048;
+const DEFAULT_HISTOGRAM_STEP_MILLIS: u64 = 50;
+
+pub struct AtomicHistogramAccumulator {
+    capacity: usize,
+    step_millis: u64,
+    buckets: Vec<AtomicU64>,
+}
+
+impl Default for AtomicHistogramAccumulator {
+    fn default() -> AtomicHistogramAccumulator {
+        AtomicHistogramAccumulator::new(DEFAULT_HISTOGRAM_CAPACITY, DEFAULT_HISTOGRAM_STEP_MILLIS)
+    }
+}
+
+impl AtomicHistogramAccumulator {
+    pub fn new(size: usize, step: u64) -> AtomicHistogramAccumulator {
+        let mut buf = Vec::with_capacity(size);
+        for _i in 0..size {
+            buf.push(AtomicU64::new(0));
+        }
+        Self {
+            capacity: size,
+            step_millis: step,
+            buckets: buf,
+        }
+    }
+
+    pub fn snapshot(&self) -> AtomicHistogramSnapshot {
+        let mut buf = Vec::with_capacity(self.capacity);
+        for i in 0..self.capacity {
+            buf.push(self.buckets[i].load(Ordering::Relaxed));
+        }
+        AtomicHistogramSnapshot {
+            capacity: self.capacity,
+            step_millis: self.step_millis,
+            buckets: buf,
+        }
+    }
+
+    fn get_bucket_num(&self, latency: u64) -> u64 {
+        let bucket_num = latency / self.step_millis;
+        if bucket_num >= self.capacity as u64 - 2 {
+            return self.capacity as u64 - 1;
+        }
+        bucket_num
+    }
+
+    pub fn record_latency_ms(&self, latency_ms: u64, num_requests: u64) {
+        let bucket_num = self.get_bucket_num(latency_ms);
+        self.buckets[bucket_num as usize].fetch_add(num_requests as u64, Ordering::Relaxed);
+    }
+}
+
+#[derive(Debug)]
+pub struct AtomicHistogramSnapshot {
+    capacity: usize,
+    step_millis: u64,
+    buckets: Vec<u64>,
+}
+
+impl Default for AtomicHistogramSnapshot {
+    fn default() -> AtomicHistogramSnapshot {
+        AtomicHistogramAccumulator::default().snapshot()
+    }
+}
+
+impl Sub for &AtomicHistogramSnapshot {
+    type Output = AtomicHistogramSnapshot;
+
+    fn sub(self, other: &AtomicHistogramSnapshot) -> AtomicHistogramSnapshot {
+        assert_eq!(
+            self.buckets.len(),
+            other.buckets.len(),
+            "Histogram snapshots must have same size, prev: {}, cur: {}",
+            self.buckets.len(),
+            other.buckets.len()
+        );
+        let mut buf = Vec::with_capacity(self.capacity);
+        for i in 0..self.buckets.len() {
+            buf.push(self.buckets[i] - other.buckets[i]);
+        }
+        AtomicHistogramSnapshot {
+            capacity: self.capacity,
+            step_millis: self.step_millis,
+            buckets: buf,
+        }
+    }
+}
+
+impl AtomicHistogramSnapshot {
+    pub fn cal_percentile_latency(
+        &self,
+        numerator: u64,
+        denominator: u64,
+        committed: u64,
```

Comment:
Do you want to just calculate this number by summing across all `snapshot.buckets`?

Reply:
if we can pass in the totally number which already have from outside, I prefer to save one operation here?

Reply:
I understand preference to save some time, but really more convenient API outweights perf in this case, in my opinion.

Reply:
for generic usage, yes, I agree, will update

Reply:
I may also change some fn name such as ```record_latency_ms``` to a generic name like ```record_data_point```